upup/pkg/fi/cloudup/openstacktasks/sshkey.go

```diff
@@ -0,0 +1,125 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openstacktasks
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/golang/glog"
+	"github.com/gophercloud/gophercloud/openstack/compute/v2/extensions/keypairs"
+
+	"k8s.io/kops/upup/pkg/fi"
+	"k8s.io/kops/upup/pkg/fi/cloudup/openstack"
+)
+
+//go:generate fitask -type=SSHKey
+type SSHKey struct {
+	Name      *string
+	Lifecycle *fi.Lifecycle
+
+	PublicKey *fi.ResourceHolder
+
+	KeyFingerprint *string
+}
+
+var _ fi.CompareWithID = &SSHKey{}
+
+func (e *SSHKey) CompareWithID() *string {
+	return e.Name
+}
+
+func (e *SSHKey) Find(c *fi.Context) (*SSHKey, error) {
+	cloud := c.Cloud.(openstack.OpenstackCloud)
+	rs, err := cloud.ListKeypair(openstackKeyPairName(fi.StringValue(e.Name)))
+	if err != nil {
+		return nil, err
+	}
+	if rs == nil {
+		return nil, nil
+	}
+	actual := &SSHKey{
+		Name:           e.Name,
+		KeyFingerprint: fi.String(rs.Fingerprint),
+	}
+
+	// Avoid spurious changes
+	if fi.StringValue(actual.KeyFingerprint) == fi.StringValue(e.KeyFingerprint) {
+		glog.V(2).Infof("SSH key fingerprints match; assuming public keys match")
+		actual.PublicKey = e.PublicKey
+	} else {
+		glog.V(2).Infof("Computed SSH key fingerprint mismatch: %q %q", fi.StringValue(e.KeyFingerprint), fi.StringValue(actual.KeyFingerprint))
+	}
+	actual.Lifecycle = e.Lifecycle
+	return actual, nil
+}
+
+func (e *SSHKey) Run(c *fi.Context) error {
+	return fi.DefaultDeltaRunMethod(e, c)
+}
+
+func (s *SSHKey) CheckChanges(a, e, changes *SSHKey) error {
+	if a == nil {
+		if e.Name == nil {
+			return fi.RequiredField("Name")
+		}
+	} else {
+		if changes.Name != nil {
+			return fi.CannotChangeField("Name")
+		}
+		if changes.KeyFingerprint != nil {
+			return fi.CannotChangeField("KeyFingerprint")
+		}
+	}
+	return nil
+}
+
+func openstackKeyPairName(org string) string {
+	name := strings.Replace(org, ".", "-", -1)
+	name = strings.Replace(name, ":", "_", -1)
+	return name
+}
+
+func (_ *SSHKey) RenderOpenstack(t *openstack.OpenstackAPITarget, a, e, changes *SSHKey) error {
+	if a == nil {
+		glog.V(2).Infof("Creating Keypair with name:%q", fi.StringValue(e.Name))
+
+		opt := keypairs.CreateOpts{
+			Name: openstackKeyPairName(fi.StringValue(e.Name)),
```

Comment:
You could probably just create the name (in the model) with a valid Name for openstack - the Name is just the identifier that you use to find the SSH Key again.  For AWS, we embed the fingerprint into the Name because SSH Keys on AWS are immutable.  Not sure if that applies on OpenStack